Completion block any input

### Summary

When using the Ctrl+n/p completion menu only vue files(volar lsp), sometimes the helix (any input) is blocked for a few seconds.



### Reproduction Steps

1. Open vue file in helix
2. Use completion(Ctrl+n/p) 
[record.webm](https://user-images.githubusercontent.com/6105977/201409055-b5b59405-73d6-4841-a996-9a0c4ecf82e1.webm)

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>
[helix-log.log](https://github.com/helix-editor/helix/files/9992674/helix-log.log)
</details>


### Platform

Linux

### Terminal Emulator

kitty 0.26.5 

### Helix Version

22.08.1-428-g5b73c8c7

Have the same problems, it looks like cause by blocking request of  [`completion/resolve`](https://github.com/helix-editor/helix/blob/master/helix-term/src/ui/completion.rs#L258).
below is the log, lsp timeout set by 5s
- 1. `completion/resolve` request sent
- 2. `workspace/configuration` request sent from server, and before the server response to `completion/resolve` request, 
       it  wait for response of  `workspace/configuration`, that result occurs dead lock
```
2022-11-14T20:58:03.863 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":6,"line":128},"start":{"character":4,"line":128}},"text":"tagID"}],"textDocument":{"uri":"file:///home/sun/example/webapp/pages/c/tags/_id.vue","version":4}}}
2022-11-14T20:58:03.868 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":7,"method":"workspace/configuration","params":{"items":[{"section":"volar.diagnostics.delay"}]}}
2022-11-14T20:58:03.886 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","result":[null],"id":7}
2022-11-14T20:58:04.262 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"completionItem/resolve","params":{"commitCharacters":[".",",",";","("],"data":{"originalItem":{"commitCharacters":[".",",",";","("],"data":{"fileName":"/home/sun/example/webapp/pages/c/tags/_id.vue.ts","offset":2862,"originalItem":{"kind":"const","kindModifiers":"","name":"tagID","sortText":"11"},"uri":"file:///home/sun/example/webapp/pages/c/tags/_id.vue.ts"},"insertTextFormat":1,"kind":6,"label":"tagID","sortText":"11","textEdit":{"insert":{"end":{"character":6,"line":86},"start":{"character":4,"line":86}},"newText":"tagID","replace":{"end":{"character":6,"line":86},"start":{"character":4,"line":86}}}},"pluginId":9,"sourceMap":{"embeddedDocumentUri":"file:///home/sun/example/webapp/pages/c/tags/_id.vue.ts"},"uri":"file:///home/sun/example/webapp/pages/c/tags/_id.vue"},"insertTextFormat":1,"kind":6,"label":"tagID","sortText":"11","textEdit":{"insert":{"end":{"character":6,"line":128},"start":{"character":4,"line":128}},"newText":"tagID","replace":{"end":{"character":6,"line":128},"start":{"character":4,"line":128}}}},"id":3}
2022-11-14T20:58:04.436 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/sun/example/webapp/pages/c/tags/_id.vue","diagnostics":[{"range":{"start":{"line":23,"character":21},"end":{"line":23,"character":29}},"severity":1,"source":"ts","code":2531,"message":"Object is possibly 'null'.","data":{"version":4}},{"range":{"start":{"line":128,"character":4},"end":{"line":128,"character":6}},"severity":1,"source":"ts","code":2304,"message":"Cannot find name 'ta'.","data":{"version":4}}],"version":4}}
2022-11-14T20:58:04.438 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":8,"method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///home/sun/example/webapp/pages/c/tags/_id.vue.ts","section":"typescript.format"}]}}
2022-11-14T20:58:04.438 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":9,"method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///home/sun/example/webapp/pages/c/tags/_id.vue.ts","section":"typescript"}]}}
2022-11-14T20:58:04.441 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/sun/example/webapp/pages/c/tags/_id.vue","diagnostics":[{"range":{"start":{"line":23,"character":21},"end":{"line":23,"character":29}},"severity":1,"source":"ts","code":2531,"message":"Object is possibly 'null'.","data":{"version":4}},{"range":{"start":{"line":128,"character":4},"end":{"line":128,"character":6}},"severity":1,"source":"ts","code":2304,"message":"Cannot find name 'ta'.","data":{"version":4}}],"version":4}}
2022-11-14T20:58:04.442 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/sun/example/webapp/pages/c/tags/_id.vue","diagnostics":[{"range":{"start":{"line":23,"character":21},"end":{"line":23,"character":29}},"severity":1,"source":"ts","code":2531,"message":"Object is possibly 'null'.","data":{"version":4}}],"version":4}}
2022-11-14T20:58:04.442 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/sun/example/webapp/pages/c/tags/_id.vue","diagnostics":[{"range":{"start":{"line":23,"character":21},"end":{"line":23,"character":29}},"severity":1,"source":"ts","code":2531,"message":"Object is possibly 'null'.","data":{"version":4}}],"version":4}}
2022-11-14T20:58:09.264 helix_term::ui::completion [ERROR] execute LSP command: request timed out
2022-11-14T20:58:09.278 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","result":[null],"id":8}
2022-11-14T20:58:09.279 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","result":[null],"id":9}
2022-11-14T20:58:09.279 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":10,"method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///home/sun/example/webapp/pages/c/tags/_id.vue.ts","section":"typescript.preferences"}]}}
2022-11-14T20:58:09.280 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","result":[null],"id":10}
2022-11-14T20:58:09.330 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":11,"method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///home/sun/example/webapp/pages/c/tags/_id.vue.ts","section":"typescript.suggest.completeFunctionCalls"}]}}
2022-11-14T20:58:09.331 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","result":[null],"id":11}
2022-11-14T20:58:09.332 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":3,"result":{"commitCharacters":[".",",",";","("],"data":{"fileName":"/home/sun/example/webapp/pages/c/tags/_id.vue.ts","offset":2862,"originalItem":{"kind":"const","kindModifiers":"","name":"tagID","sortText":"11"},"uri":"file:///home/sun/example/webapp/pages/c/tags/_id.vue.ts"},"insertTextFormat":1,"kind":6,"label":"tagID","sortText":"11","textEdit":{"insert":{"start":{"line":128,"character":4},"end":{"line":128,"character":6}},"newText":"tagID","replace":{"start":{"line":128,"character":4},"end":{"line":128,"character":6}}},"detail":"const tagID: string","documentation":{"kind":"markdown","value":""}}}
2022-11-14T20:58:09.332 helix_lsp::transport [INFO] <- {"commitCharacters":[".",",",";","("],"data":{"fileName":"/home/sun/example/webapp/pages/c/tags/_id.vue.ts","offset":2862,"originalItem":{"kind":"const","kindModifiers":"","name":"tagID","sortText":"11"},"uri":"file:///home/sun/example/webapp/pages/c/tags/_id.vue.ts"},"detail":"const tagID: string","documentation":{"kind":"markdown","value":""},"insertTextFormat":1,"kind":6,"label":"tagID","sortText":"11","textEdit":{"insert":{"end":{"character":6,"line":128},"start":{"character":4,"line":128}},"newText":"tagID","replace":{"end":{"character":6,"line":128},"start":{"character":4,"line":128}}}}
2022-11-14T20:58:09.332 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(3)), original request likely timed out
```

--- a/helix-lsp/src/client.rs
+++ b/helix-lsp/src/client.rs
@@ -650,12 +650,11 @@ impl Client {
         self.call::<lsp::request::Completion>(params)
     }
 
-    pub async fn resolve_completion_item(
+    pub fn resolve_completion_item(
         &self,
         completion_item: lsp::CompletionItem,
-    ) -> Result<lsp::CompletionItem> {
-        self.request::<lsp::request::ResolveCompletionItem>(completion_item)
-            .await
+    ) -> impl Future<Output = Result<Value>> {
+        self.call::<lsp::request::ResolveCompletionItem>(completion_item)
     }
 
     pub fn text_document_signature_help(

--- a/helix-term/src/ui/completion.rs
+++ b/helix-term/src/ui/completion.rs
@@ -257,7 +257,7 @@ impl Completion {
         let future = language_server.resolve_completion_item(completion_item);
         let response = helix_lsp::block_on(future);
         match response {
-            Ok(completion_item) => Some(completion_item),
+            Ok(value) => serde_json::from_value(value).ok(),
             Err(err) => {
                 log::error!("execute LSP command: {}", err);
                 None
@@ -304,6 +304,12 @@ impl Completion {
         self.popup.contents().is_empty()
     }
 
+    fn replace_item(&mut self, old_item: lsp::CompletionItem, new_item: lsp::CompletionItem) {
+        self.popup.contents_mut().replace_option(old_item, new_item);
+    }
+
+    /// Asynchronously requests that the currently selection completion item is
+    /// resolved through LSP `completionItem/resolve`.
     pub fn ensure_item_resolved(&mut self, cx: &mut commands::Context) -> bool {
         // > If computing full completion items is expensive, servers can additionally provide a
         // > handler for the completion item resolve request. ...
@@ -313,16 +319,38 @@ impl Completion {
         // > 'completionItem/resolve' request is sent with the selected completion item as a parameter.
         // > The returned completion item should have the documentation property filled in.
         // https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion
-        match self.popup.contents_mut().selection_mut() {
-            Some(item) if item.documentation.is_none() => {
-                let doc = doc!(cx.editor);
-                if let Some(resolved_item) = Self::resolve_completion_item(doc, item.clone()) {
-                    *item = resolved_item;
+        let current_item = match self.popup.contents().selection() {
+            Some(item) if item.documentation.is_none() => item.clone(),
+            _ => return false,
+        };
+
+        let language_server = match doc!(cx.editor).language_server() {
+            Some(language_server) => language_server,
+            None => return false,
+        };
+
+        // This method should not block the compositor so we handle the response asynchronously.
+        let future = language_server.resolve_completion_item(current_item.clone());
+
+        cx.callback(
+            future,
+            move |_editor, compositor, response: Option<lsp::CompletionItem>| {
+                let resolved_item = match response {
+                    Some(item) => item,
+                    None => return,
+                };
+
+                if let Some(completion) = &mut compositor
+                    .find::<crate::ui::EditorView>()
+                    .unwrap()
+                    .completion
+                {
+                    completion.replace_item(current_item, resolved_item);
                 }
-                true
-            }
-            _ => false,
-        }
+            },
+        );
+
+        true
     }
 }
 

--- a/helix-term/src/ui/menu.rs
+++ b/helix-term/src/ui/menu.rs
@@ -224,6 +224,17 @@ impl<T: Item> Menu<T> {
     }
 }
 
+impl<T: Item + PartialEq> Menu<T> {
+    pub fn replace_option(&mut self, old_option: T, new_option: T) {
+        for option in &mut self.options {
+            if old_option == *option {
+                *option = new_option;
+                break;
+            }
+        }
+    }
+}
+
 use super::PromptEvent as MenuEvent;
 
 impl<T: Item + 'static> Component for Menu<T> {

